src/batch/src/executor/join/lookup_join.rs

```diff
@@ -162,9 +165,30 @@ impl<C: BatchTaskContext> ProbeSideSourceBuilder for ProbeSideSource<C> {
     /// hash map, along with the scan range's virtual node.
     fn add_scan_range(&mut self, key_scalar_impls: &[ScalarImpl]) -> Result<()> {
         let mut scan_range = ScanRange::full_table_scan();
-        key_scalar_impls
+
+        for ((scalar_impl, build_type), probe_type) in key_scalar_impls
             .iter()
-            .for_each(|scalar_impl| scan_range.eq_conds.push(scalar_impl.clone()));
+            .zip_eq(self.build_side_key_types.iter())
+            .zip_eq(self.probe_side_key_types.iter())
+        {
+            let scalar_impl = if probe_type == build_type {
+                scalar_impl.clone()
+            } else {
+                let cast_expr = new_unary_expr(
+                    Type::Cast,
+                    probe_type.clone(),
+                    Box::new(LiteralExpression::new(
+                        build_type.clone(),
+                        Some(scalar_impl.clone()),
+                    )),
+                )?;
+
+                let datum = cast_expr.eval_row(&Row(vec![]))?;
+                datum.unwrap()
```

Comment:
Does it mean that `datum` cannot be `None`?

Reply:
Yep

Reply:
Actually a panic occurs if `null` is inserted into the table but it's an issue caused by the other parts of the code. I think I can fix this behavior in a different PR.

Reply:
Can add the `mergify/can-merge` to let this PR automatically merge with the main branch.